lib/tasks/pipeline.js

```diff
@@ -38,58 +36,83 @@ module.exports = Task.extend({
-    this.configPath        = 'config/environment.js';
+    this.appConfigPath        = 'config/environment.js';
 
     if (!existsSync(this.deployConfigPath)) {
       throw new SilentError('Deploy config does not exist at `' + this.deployConfigPath + '`');
     }
 
-    if (!existsSync(this.configPath)) {
-      throw new SilentError('App config does not exist at `' + this.configPath + '`');
+    if (!existsSync(this.appConfigPath)) {
+      throw new SilentError('App config does not exist at `' + this.appConfigPath + '`');
     }
+  },
 
-    this._deployConfig = require(path.join(this.project.root, this.deployConfigPath))(this.deployEnvironment);
-
-    var buildEnv = (this._deployConfig.build && this._deployConfig.build.buildEnv) || 'production';
-
-    this._appConfig = require(path.join(this.project.root, this.configPath))(buildEnv);
+  setup: function(){
+    var self = this;
+    return this._readConfig().then(function(configs){
+      self.project.addons.forEach(function(addon){
+        self._registerPipelineHooks(addon, configs.deployConfig.plugins);
+      });
+      return configs;
+    });
   },
 
   run: function(hooks) {
     var pipeline = this._pipeline;
-    var context = {
-      ui: this.ui,
-      project: this.project,
-      config: this._deployConfig,
-      appConfig: this._appConfig
-    };
-
-    return pipeline.execute({
-      deployment: context
+    var ui = this.ui;
+    var project = this.project;
+    var self = this;
+    return this.setup().then(function(configs){
+      var context = {
+        ui: ui,
+        project: project,
+        config: configs.deployConfig,
+        appConfig: configs.appConfig
+      };
+      return pipeline.execute({
+        deployment: context
+      });
     });
   },
 
-  _registerPipelineHooks: function(addon) {
+  _readConfig: function() {
+    var project = this.project;
+    var appConfigPath = this.appConfigPath;
+    var deployConfigFn = require(path.join(project.root, this.deployConfigPath));
+    return Promise.resolve(deployConfigFn(this.deployEnvironment)).then(function(deployConfig){
+      var buildEnv = (deployConfig.build && deployConfig.build.buildEnv) || 'production';
+      var appConfigFn = require(path.join(project.root, appConfigPath));
+      return {
+        deployConfig: deployConfig,
+        appConfig: appConfigFn(buildEnv)
+      };
+    });
+  },
+
+  _registerPipelineHooks: function(addon, configuredPluginList) {
     var isValidDeployPlugin = this._isValidDeployPlugin(addon);
+    if (!isValidDeployPlugin) { return; }
 
-    if (isValidDeployPlugin) {
-      var pluginNameRegex = /^(ember\-cli\-deploy\-)(.*)$/;
-      var name            = addon.name.match(pluginNameRegex)[2];
-      var pluginHooks     = addon.createDeployPlugin({name: name});
+    var pluginNameRegex = /^(ember\-cli\-deploy\-)(.*)$/;
+    var name            = addon.name.match(pluginNameRegex)[2];
+    if (configuredPluginList && configuredPluginList.indexOf(name) === -1) {
```

Comment:
@lukemelia , Am I right in saying that this currently expects the `configuredPluginList` to be an array of the actual plugin in names, eg `['build', 'tag', 's3', 'redis']`.....as opposed to the goal of the advanced config being `['build', 'tag', 's3:index', 's3:assets']` ? I'm guessing this is a step in that direction yeh?